File: client/src/main/java/au/com/codeka/warworlds/client/store/StarCursor.java
package au.com.codeka.warworlds.client.store;

import android.database.Cursor;

import java.io.IOException;
import java.util.Iterator;

import javax.annotation.Nullable;

import au.com.codeka.warworlds.common.proto.Star;

/**
 * A "cursor" that's used to scroll through the results of querying stars.
 */
public class StarCursor implements Iterable<Star>, AutoCloseable {
  private final Cursor cursor;

  public StarCursor(Cursor cursor) {
    this.cursor = cursor;
  }

  public int getSize() {
    return cursor.getCount();
  }

  /**
   * Gets the value at the given position, can return null if the position is invalid or the star
   * could not be deserialized.
   */
  @Nullable
  public Star getValue(int position) {
    cursor.moveToPosition(position);
    try {
      return Star.ADAPTER.decode(cursor.getBlob(0));
    } catch (IOException e) {
      return null;
    }
  }

  @Override
  public void close() {
    cursor.close();
  }

  @Override
  public Iterator<Star> iterator() {
    return new StarIterator(cursor);
  }

  private static class StarIterator implements Iterator<Star> {
    private final Cursor cursor;
    private boolean hasValue;

    public StarIterator(Cursor cursor) {
      this.cursor = cursor;
      hasValue = cursor.moveToFirst();
    }

    @Override
    public boolean hasNext() {
      return hasValue;
    }

    @Override
    public Star next() {
      try {
        Star star = Star.ADAPTER.decode(cursor.getBlob(0));
        hasValue = cursor.moveToNext();
        return star;
      } catch (IOException e) {
        return null;
      }
    }
  }
}

File: client/src/main/kotlin/au/com/codeka/warworlds/client/store/StarCursor.kt
package au.com.codeka.warworlds.client.store

import android.database.Cursor
import au.com.codeka.warworlds.common.proto.Star
import java.io.IOException

/**
 * A "cursor" that's used to scroll through the results of querying stars.
 */
class StarCursor(private val cursor: Cursor) : Iterable<Star?>, AutoCloseable {
  val size: Int
    get() = cursor.count

  /**
   * Gets the value at the given position, can return null if the position is invalid or the star
   * could not be deserialized.
   */
  fun getValue(position: Int): Star {
    cursor.moveToPosition(position)
    return Star.ADAPTER.decode(cursor.getBlob(0))
  }

  override fun close() {
    cursor.close()
  }

  override fun iterator(): Iterator<Star> {
    return StarIterator(cursor)
  }

  private class StarIterator(private val cursor: Cursor) : Iterator<Star> {
    private var hasValue: Boolean
    override fun hasNext(): Boolean {
      return hasValue
    }

    override fun next(): Star {
      return try {
        val star = Star.ADAPTER.decode(cursor.getBlob(0))
        hasValue = cursor.moveToNext()
        star
      } catch (e: IOException) {
        throw NoSuchElementException()
      }
    }

    init {
      hasValue = cursor.moveToFirst()
    }
  }

}